core/network/src/protocol.rs

```diff
@@ -393,19 +393,8 @@ impl<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> Protocol<B, S, H> {
 		trace!(target: "sync", "BlockResponse {} from {} with {} blocks{}",
 			response.id, peer, response.blocks.len(), blocks_range);
 
-		// import_queue.import_blocks also acquires sync.write();
-		// Break the cycle by doing these separately from the outside;
-		let new_blocks = {
-			let mut sync = self.sync.write();
-			sync.on_block_data(&mut ProtocolContext::new(&self.context_data, io), peer, request, response)
-		};
-
-		if let Some((origin, new_blocks)) = new_blocks {
-			let import_queue = self.sync.read().import_queue();
-			import_queue.import_blocks(origin, new_blocks);
-		}
-
-
+		let mut sync = self.sync.write();
+		sync.on_block_data(&mut ProtocolContext::new(&self.context_data, io), peer, request, response)
```

Comment:
Ok, so, like the comment above read, this is indeed deadlocking :)

Could be something that goes away using channels, which I'll try later, so leaving this for now...